### DISCUSSION TITLE
Difference between Firebase Application Project and Firebase Function Application Project

### DISCUSSION THREAD

**Discussion by j1mmie** [simondotm/nx-firebase#178] [General]
Hi @simondotm - I've been using nx-firebase for about a year now, but I was on the 1.0.0 release. I've recently upgraded to 2.1.2 and I'm having trouble migrating my existing functions app over.

In 1.0.0, I had 3 apps:

- apps/functions
- apps/game
- apps/auth

And 2 libs:

- libs/core
- libs/shared

`game` is a game server, `auth` is an OAuth web server, and functions were my Firebase functions. This all worked fine and well.

[Reading the documentation for 2.1.2](https://github.com/simondotm/nx-firebase/blob/main/docs/quick-start.md), I see two concepts that I am having trouble understanding; Firebase Application Project and Firebase Function Application Project.

How do these concepts map to my 1.0.0 folder structure? I believe my `functions` folder would be akin to a Firebase Function Application Project. Do I need a Firebase Application Project?

The documentation is great but I'm just having trouble understanding the big picture. I have a general idea of the things I *can* do with nx-firebase but I'm not sure why I should do them, what purpose they serve, etc

**Comment by j1mmie** [comment]
Just a follow up - I see some clarification here:
https://github.com/simondotm/nx-firebase/blob/main/docs/nx-firebase-applications.md

> An Nx Firebase application project is a top-level container for the various configurations for Firebase features you might wish to use such as functions, storage, firestore, and real time database.

So a Firebase Application Project is intended to store configuration for a particular Firebase project, and a Firebase Function Application Project is where the actual functions go. Is that correct?

**Comment by simondotm** [comment]
Hello @j1mmie 

> So a Firebase Application Project is intended to store configuration for a particular Firebase project, and a Firebase Function Application Project is where the actual functions go. Is that correct?

Yes that is correct. For your scenario, with v2 plugin you would want to create a firebase application project in Nx, and then a firebase function project.

As a common convention you might call these v2 app projects 'firebase' and 'functions' 

Previously with v1 plugin, these would both be wrapped up in one Nx project. 

The benefit of this new structure is that by decoupling functions from the firebase root project, we can have multiple 'function' apps within a firebase project if we like, which allows for functions to be more streamlined.

Hope that helps

**Comment by j1mmie** [comment]
Thanks, that helps a lot. I actually named my two apps exactly that. I like the new structure, makes a lot of sense.
